### ISSUE TITLE
MAGMA on Perlmutter cannot find cublas/cusparse

### ISSUE THREAD

**Issue by pghysels** [xsdk-project/xsdk-issues#195]
On Perlmutter, the CUDA math libs are in a separate folder from the core CUDA libs:
`/opt/nvidia/hpc_sdk/Linux_x86_64/22.5/math_libs/11.7/lib64` vs `/opt/nvidia/hpc_sdk/Linux_x86_64/22.5/cuda/11.7/lib64`.

MAGMA 2.6.2 cannot find `cuBLAS` or `cuSPARSE`.

This is fixed in MAGMA@master, which uses `find_package( CUDAToolkit )` and links to the CMake targets `CUDA::cublas` and `CUDA::cusparse`, whereas MAGMA 2.6.2 instead uses the deprecated `find_package( CUDA )`.

**Comment by pghysels** [comment]
This is also a problem for Tasmanian

**Comment by pghysels** [comment]
heffte cannot find cufft

**Comment by pghysels** [comment]
Attached is a patch for heffte, which works for me on Perlmutter.
Should this be added in spack, or can it go in a new heffte release?
[cudatoolkit.patch.txt](https://github.com/xsdk-project/xsdk-issues/files/9840993/cudatoolkit.patch.txt)

**Comment by mkstoyanov** [comment]
This makes a huge bump the minimum CMake right before we tag the release. I want to update CMake, but not right now.

I'll add the workaround, will enable the newer CUDA standard, but will keep the CMake requirements. You'll only need to use newer CMake on Perlmutter.

Thanks for reporting the problem.

**Comment by pghysels** [comment]
I understand your concern.
(I should have used a minimum CMake 3.17 instead of 3.18, FindCUDAToolkit was introduced in CMake 3.17).

My Ubuntu and Fedora machines both have CMake 3.22.
Cori and Perlmutter use CMake 3.17 by default but you can easily switch to 3.22 by reloading the module.

**Comment by mkstoyanov** [comment]
The plan was to update the CMake version right after the release, I have to change that now.

Ubuntu 20.04 still uses CMake 3.16 as a default and ROCm has no support for Ubuntu 22.04 yet. Requiring CMake 3.17 means that we have to manually handle CMake installations in the CI and will affect some users. Let's hope the optional dependence will be enough.

**Comment by balay** [comment]
Wrt xsdk - some pkgs already have higher cmake requirement - so a newer cmake is build and used [if not preinstalled] for all pkgs in xsdk install.

superlu-dist:
```
    depends_on("cmake@3.18.1:", type="build", when="@7.1.0:")
```
amrex:
```
    depends_on("cmake@3.17:", type="build", when="@22.06:")
    depends_on("cmake@3.20:", type="build", when="+rocm")
```
sundials:
```
    depends_on("cmake@3.12:", type="build")
    depends_on("cmake@3.18:", when="+cuda", type="build")
```

**Comment by mkstoyanov** [comment]
That's fair @balay but not all users are going through spack.

**Comment by balay** [comment]
Sure - just wanted to point that out.. [and for some reason different cmake requirements for different usages...]

Also presumably supporting builds on  perlmutter is independent of xsdk - so you would decide on the appropriate way to deal with this issue

If it were only for xsdk/spack - you can always add in that patch in spack [and not in mama release]

**Comment by mkstoyanov** [comment]
Just updated Tasmanian and heFFTe, `spack install tasmanian@develop+cuda` should work fine now.

**Comment by pghysels** [comment]
Thanks, that works.

**Comment by mkstoyanov** [comment]
I guess we can close the issue.

**Comment by pghysels** [comment]
These issues are now fixed now in magma@master, heffte@develop and tasmanian@develop.
But I'm not sure how this will be handled for the xsdk@0.8.0 release?
I'll leave that to @balay

**Comment by mkstoyanov** [comment]
heffte and tasmanian develop variants were converted to releases

**Comment by pghysels** [comment]
For MAGMA, we could use this commit as a patch
https://bitbucket.org/icl/magma/commits/4c0c4df0cab79cf19bdb8917a56788f744495fca

I reopened this for MAGMA. Tasmanian and heffte are fine now.
This issue was for all 3 packages, because the issue is the same.

**Comment by mkstoyanov** [comment]
@pghysels It's OK to reopen this for MAGMA, but make sure to tag some of the MAGMA developers. I don't work on MAGMA, just tasmanian and heffte.

**Comment by balay** [comment]
> @pghysels It's OK to reopen this for MAGMA, but make sure to tag some of the MAGMA developers. I don't work on MAGMA, just tasmanian and heffte.

cc: @luszczek @hartwiganzt @stomov

**Comment by stomov** [comment]
Thanks for reporting this. We had it indeed fixed in master - this will be available in the new magma 2.7 release scheduled for tomorrow. I will comment here again when we do it.

**Comment by stomov** [comment]
@pghysels MAGMA 2.7 is now released with a fix for this issue, and @balay already updated it in spack (pull request https://github.com/spack/spack/pull/33814).
